### PR TITLE
Implement keyword recommenders for `allows ref struct` constraint

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders;
+
+internal class AllowsKeywordRecommender : AbstractSyntacticSingleKeywordRecommender
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs:12]:
```suggestion
internal sealed class AllowsKeywordRecommender : AbstractSyntacticSingleKeywordRecommender
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        var token = context.TargetToken;
+
+        if (token.Kind() == SyntaxKind.CommaToken &&
+            token.Parent is TypeParameterConstraintClauseSyntax constraintClause)
+        {
+            if (!constraintClause.Constraints
+                    .OfType<ClassOrStructConstraintSyntax>()
+                    .Any(c => c.ClassOrStructKeyword.Kind() == SyntaxKind.ClassKeyword))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs:45]: note: we'd be fine with it still showing up.  that way someone can type it, but get a good erro rmessage.  but we're fine either way on this.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/StructKeywordRecommender.cs [hunk 3]
```diff
@@ -39,6 +39,34 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
...
+        //    where T : class, allows ref |
+        //    where T : new(), allows ref |
+        //    where T : Goo, allows ref |
+
+        var token = context.TargetToken;
+
+        if (token.Kind() == SyntaxKind.RefKeyword &&
+            token.Parent is RefStructConstraintSyntax refStructConstraint && refStructConstraint.RefKeyword == token &&
+            refStructConstraint.Parent is AllowsConstraintClauseSyntax allowsClause &&
+            allowsClause.Parent is TypeParameterConstraintClauseSyntax)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/StructKeywordRecommender.cs:65]: you can trim off this last check.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/RefKeywordRecommender.cs [hunk 4]
```diff
@@ -85,7 +87,35 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
...
+        var token = context.TargetToken;
+
+        if (token.Kind() == SyntaxKind.IdentifierToken && SyntaxFacts.GetContextualKeywordKind((string)token.Value!) == SyntaxKind.AllowsKeyword &&
+            token.Parent is IdentifierNameSyntax identifier && identifier.Identifier == token &&
+            identifier.Parent is TypeConstraintSyntax typeConstraint && typeConstraint.Type == identifier &&
+            typeConstraint.Parent is TypeParameterConstraintClauseSyntax constraintClause)
+        {
+            if (!constraintClause.Constraints
+                    .OfType<ClassOrStructConstraintSyntax>()
+                    .Any(c => c.ClassOrStructKeyword.Kind() == SyntaxKind.ClassKeyword))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/RefKeywordRecommender.cs:112]: we'd be fine with this being dropped.

File: src/EditorFeatures/CSharpTest2/Recommendations/StructKeywordRecommenderTests.cs [hunk 5]
```diff
@@ -521,5 +521,168 @@ public async Task TestNotAfterRecord_SealedModifier()
...
+                      where T : allows ref $$
+                      where U : T
+                """);
+        }
+
+        [Fact]
+        public async Task TestAfterAllowsRefAfterClassTypeParameterConstraint()
+        {
+            await VerifyKeywordAsync(
+@"class C<T> where T : class, allows ref $$");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/StructKeywordRecommenderTests.cs:570]: could these all be raw strings plz.  or indent 4 from `await`.  we're trying to move away entirely from tests that are 0-column aligned.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs:45]: > note: we'd be fine with it still showing up.  that way someone can type it, but get a good erro rmessage.  but we're fine either way on this. 

Following what `new` recommender does, doesn't offer if `struct` is present

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm with minor nits, simplification suggestions, and requests about test formatting.

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm with minor nits, simplification suggestions, and requests about test formatting.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AllowsKeywordRecommender.cs:45]: wfm.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/RefKeywordRecommender.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/RefKeywordRecommender.cs:112]: > we'd be fine with this being dropped.  

Following what we do for `new` constraint

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs [hunk 6]
```diff
@@ -5979,5 +5979,86 @@ public void TestNormalizeParseExpressionLiteralCharacter(string expression)
...
+                :
+                allows
+                ref
+                struct
+                ;
+                """, """
+                class C1<T>
+                  where T : allows ref struct;
+                class C2<T, S>
+                  where T : allows ref struct , where S : struct;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:6003]: The space before the comma looks unexpected. Shouldn't it be:
```suggestion
                  where T : allows ref struct, where S : struct;
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/StructKeywordRecommender.cs [hunk 7]
```diff
@@ -39,6 +39,34 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
...
                 cancellationToken: cancellationToken) ||
             context.IsRecordDeclarationContext(s_validModifiers, cancellationToken) ||
-            syntaxTree.IsTypeParameterConstraintStartContext(position, context.LeftToken);
+            IsAllowsRefStructConstraintContext(context);
+    }
+
+    private static bool IsAllowsRefStructConstraintContext(CSharpSyntaxContext context)
+    {
+        //    where T : |
+        if (context.SyntaxTree.IsTypeParameterConstraintStartContext(context.Position, context.LeftToken))
```
Comment (jjonescz) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/StructKeywordRecommender.cs:48]: I wouldn't expect this check to be inside method `IsAllowsRefStructConstraintContext`, consider returning it where it was previously.

Comment (AlekseyTs) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/StructKeywordRecommender.cs:48]: I'll rename the method

File: src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNormalizerTests.cs:6003]: > The space before the comma looks unexpected. 

It is a syntax error, a comma cannot follow `allows ref struct` constraint

**Review by jjonescz — APPROVED** [review]

**Pull request merged**